### ISSUE TITLE
Astropy returns error when imported by Python 3.4 within an empty git repo

### ISSUE THREAD

**Issue by wcleveland** [astropy/astropy#3475]
Python version: 3.4
Astropy version: 0.4.4

**Problem:**
 When the project directory (really working directory) is under git version control, you can not import the astropy library. After removing the '.git' directory (eliminating the version control), the library can be imported without errors.

**Steps to recreate the problem:**
```
(install astropy via pip)
$ mkdir test
$ cd test
$ git init
$ echo "import astropy" > test.py
$ echo "print(\"Astropy version = %s\" % (astropy.__version__))" >> test.py
$ python3 test.py
(Error appears)

$ rm -rf .git
$ python3 test.py
Astropy version = 0.4.4
```

**Error message:**
```
Python 3.4.2 (default, Dec 30 2014, 18:34:07)
[GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.56)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import astropy
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/TestUser/venv/trigcat3/lib/python3.4/site-packages/astropy/__init__.py", line 23, in <module>
    from .version import version as __version__
  File "/Users/TestUser/venv/trigcat3/lib/python3.4/site-packages/astropy/version.py", line 151, in <module>
    version = update_git_devstr(_last_generated_version)
  File "/Users/TestUser/venv/trigcat3/lib/python3.4/site-packages/astropy/version.py", line 34, in update_git_devstr
    devstr = get_git_devstr(sha=True, show_warning=False, path=path)
  File "/Users/TestUser/venv/trigcat3/lib/python3.4/site-packages/astropy/version.py", line 144, in get_git_devstr
    return _decode_stdio(stdout)[:40]
  File "/Users/TestUser/venv/trigcat3/lib/python3.4/site-packages/astropy/version.py", line 17, in _decode_stdio
    text = stream.decode(stdio_encoding)
AttributeError: 'str' object has no attribute 'decode'
```

**Comment by astrofrog** [comment]
@wcleveland - could you try the same, but with ``pip install astropy --pre`` to try installing the 1.0 release candidate?

In any case, this check shouldn't be happening. I think this is a duplicate of https://github.com/astropy/astropy/issues/2561. It seems like we really should resolve this? @eteq @embray

**Comment by embray** [comment]
Yeah, I'm working on that today actually.  Something else seems odd about this though.  It's saying the `stdout` returned by `Popen.communicate` is a `str` object and not a `bytes` object like it's expected to be.  I've never encountered this before, but I don't see what it would have directly to do with #2561.

**Comment by embray** [comment]
It seems that if `universal_newlines=True` then it *will* return strings decoded using `locale.getpreferredencoding(False)`.  But that's still not what the version.py module is doing.  Is there some global variable or flag in Python that enables this functionality globally?

**Comment by embray** [comment]
Ah... *facepalm* I see.  In the `run_git` function there are a couple cases like:
```
        except OSError as e:
            if show_warning:
                warnings.warn('Error running git: ' + str(e))
            return (None, '', '')

        if p.returncode == 128:
            if show_warning:
                warnings.warn('No git repository present at {0!r}! Using '
                              'default dev version.'.format(path))
            return (p.returncode, '', '')
```
which *are* returning strings.  This error conditioning is actually occurring since it's an empty test repo, a la astropy/astropy-helpers#114.   Which does also go back to #2561 after all.

**Comment by wcleveland** [comment]
The prerelease version has the same error.

My (very untested) workaround is (beginning at line 10 of version.py):
```
def _decode_stdio(stream):
   if stream.__string__ == str:
    return stream
  ...
```

**Comment by embray** [comment]
Note, this also relates to the *specific* issue I was working on yesterday which is that the version.py included in release versions should not even contain any of the git update code.  It should be just a static file with the version strings in it, and no otherwise executable code.

**Comment by wcleveland** [comment]
I agree with your "facepalm" comment (the contents not the actual faceplate ;) )

**Comment by wcleveland** [comment]
We could close this issue and generate a new one saying that the release versions of astropy should only contain a static version.py file and not the git version functions.

**Comment by eteq** [comment]
Aha, so this only happens with an empty repo, is that right @embray? I was about to comment that I regularly import astropy sucessfully inside git repos (a la the original report in #2561), but I don't think I'd ever tried in an empty one...

Also, @embray, the actual fixes to this will all be in astropy helpers, right?  So no reason to delay RC2 of astropy for this?

**Comment by wcleveland** [comment]
If astropy can't handle a case where the user has an empty repo then I don't consider it release ready, but that's just my humble opinion. You could go ahead with RC2 if you already plan to have an RC3 anyway.

**Comment by wcleveland** [comment]
The work around of committing something (anything) to the repo prior to using astropy does work. So this issue only affects empty repos as suggested by @eteq

**Comment by eteq** [comment]
@wcleveland - I meant that if the fix belongs in astropy-helpers, which is a separate package from the astropy core, then the fix needs to be there, rather than in the astropy repo.  One of the points of astropy-helpers is that it can have releases separate from astropy for situations just like this

**Comment by wcleveland** [comment]
@eteq Understood. The only issue I have is that this problem appeared while using a release version installed via pip. Otherwise I would have closed this as a duplicate of astropy/astro-helpers#114.

Technically this is related to #2561 with the exception that the script can't run at all within an empty git repo. The #2561 is a minor issue that only affect people who check version numbers, while this issue is a show stopper at least until the first commit. At the very least, it should be reported as a known issue.

**Comment by astrofrog** [comment]
@embray - I agree with https://github.com/astropy/astropy/issues/3475#issuecomment-73739334 - just to check, do you have a solution at this point, or should one of us work on it?

**Comment by eteq** [comment]
@wcleveland - gotcha.  Now that I think about it, the scheme proposed by @embray and @astrofrog does indeed require testing in an RC because even if the code is in the helpers, the frozen version.py will end up in the source tarball.

So definitely need this for the RC...

**Comment by embray** [comment]
There is a workaround regardless which is to make the following change in `setup.py`:
```diff
diff --git a/setup.py b/setup.py
index 5cab87e..b2df7bf 100755
--- a/setup.py
+++ b/setup.py
@@ -49,7 +49,8 @@ cmdclassd = register_commands(NAME, VERSION, RELEASE)
 adjust_compiler(NAME)
 
 # Freeze build information in version.py
-generate_version_py(NAME, VERSION, RELEASE, get_debug_option(NAME))
+generate_version_py(NAME, VERSION, RELEASE, get_debug_option(NAME),
+                    uses_git=not RELEASE)
 
 # Get configuration information from all of the various subpackages.
 # See the docstring for setup_helpers.update_package_files for more
```

As I mentioned elsewhere the machinery for this was already there.  It just wasn't being used.  Meanwhile I'm making some improvements on the astropy-helpers end.  But this can be done now.

**Comment by embray** [comment]
In fact I might even release an Astropy v0.4.5 to include this change, since without it it means importing astropy takes longer since it is still trying to do stuff with git that's unnecessary.

**Comment by embray** [comment]
The fixes on astropy-helper's end are in astropy/astropy-helpers#137.  In order to call *this* issue fixed the setup.py should be updated, though we should also advance the bundled astropy-helpers version (the latter shouldn't be strictly needed but that will have to happen before the final release anyhow).

**Comment by embray** [comment]
Fixed by #3484 .

**Comment by eteq** [comment]
not yet fixed in master (or 1.0.x), so I'm reopening, but will prepare a PR shortly for that.

**Comment by wcleveland** [comment]
Thanks for the quick response.
